lib/jbuilder/jbuilder_template.rb

```diff
@@ -121,7 +123,9 @@ def array!(collection = [], *args)
     options = args.first
 
     if args.one? && _partial_options?(options)
-      partial! options.merge(collection: collection)
+      options = options.dup
+      options[:collection] = collection
+      _render_partial_with_options options
```

Comment:
Prevents mutating options on `array!` in a way that allows us to safely mutate the options in private methods to keep the other optimizations in place.